### ISSUE TITLE
Type inferring error about list comprehensions.

### ISSUE THREAD

**Issue by giyyapan** [microsoft/pyright#325]
**Describe the bug**
Please see this example.
![image](https://user-images.githubusercontent.com/3479453/68268742-5e90e200-0092-11ea-8593-39aca32bebf2.png)

The variable `a` used in `[a for a in range(10)]` is a local variable, which should not influence the variable `a` outside the expression.

The code:
```python
a = dict(m=1)
arr = [a for a in range(10)]
a["m"]
```
![image](https://user-images.githubusercontent.com/3479453/68268886-c0514c00-0092-11ea-86e4-5cc250bc0ab4.png)

Additionally, if we change the list comprehension to generator expression, there will be the same problem.
Code:
```python
a = dict(m=1)
arr = (a for a in range(10))
a["m"]
```

**VS Code extension or command-line**
VS Code extension version 1.0.83

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem as reported. I understand the underlying cause and will work on a solution.

**Comment by erictraut** [comment]
This will be fixed in the next release of pyright.

**Comment by erictraut** [comment]
This is now fixed in 1.0.84, which I just published.
